### ISSUE TITLE
Impossible to load a `Kaart`

### ISSUE THREAD

**Issue by timvisee** [gameoftrades/gameoftrades-library#1]
A `Wereld` instance needs to be created in order to effectively use the library. The constructor of the `Wereld` class requires a `Kaart` instance along with a few other parameters.

With the current code base it's only possible to specify the dimensions of `Kaart` while instantiating it, it is however not possible to specify the terrain tiles on the `Kaart` itself.

Loading a `Kaart` with any dimensions above 0 results in an exception, because the actual terrain tiles in the `Kaart` defaults to null. Loading a `Kaart` with the `0, 0` dimensions is of course possible, but those instances aren't _usable_.

The `Kaart` class (and most other library classes) are final, thus they can't be extended to add the desired implementations.

**TL;DR:** `Kaart` classes can't be instantiated with their terrain tile data.

**Comment by ctrl-alt-dev** [comment]
Did you consider using the `void setTerrein(Coordinaat c, Terrein t)` method?

**Comment by timvisee** [comment]
The `void setTerrein(...)` method is package local, thus not accessible from outside the library code base itself.

I just figured out that the `Terrein` constructor uses a `Kaart` parameter, and internally sets the proper terrain tile on the `Kaart` object.

See my example below:

```
Kaart kaart = new Kaart(1, 1);
Terrein terrein = new Terrein(kaart, new Coordinaat(0, 0), TerreinType.GRASLAND);
```

This solution however, _feels_ like a workaround to me and I'd suggest making the method you mentioned accessible. Adding an explicit `public` visibility to the `void setTerrein(...)` method would make it accessible in student projects using this library. 

Edit: Related pull request: #2

**Comment by ctrl-alt-dev** [comment]
_Good, you have found a way to composite the Terrein into the Kaart!_

Now let me explain why I will not be accepting your pull request.

Part of the exercise is to write a loader component. This component has the task of converting a given file into various domain objects which are part of the library. The library is given. 

As a design requirement for the library I need you to be able create the Kaart (and Terrein) instances and compose them but at the same time expose a safe (read-only) API for use with the algorithms.

That rules out a regular public setter which would allow modification of the map (note that the current setter is not a regular setter nor is it intended as such)

The most obvious choice would be a Kaart constructor that takes all the state (w,h,terrein). However, that would require the caller to know how to exactly encode x and y coordinates in the array. It would require me to copy the arrays to prevent modification. It requires the caller to fill the array before creating the Kaart while the map format first defines the Kaart sizes and then the Terrein. 

Not exactly ideal.

Also Terrein and Kaart are linked to each other bi-directionally (for the look-around functionality). I need some way to guarantee that all instances of Terrein link back to their Kaart and that a Kaart can not share Terrein with another Kaart. If the setter on Kaart was present, there should also be a setter on Terrein. Now there are 2 types with mutable state. 

Not good at all.

The easiest way to have mutability at construction time and immutability during runtime is to allow modification only by creating new instances. And that is exactly what the library allows you to do :)

The setter is there to allow Terrein to only modify the Kaart during its construction.

I hope you now understand the rationale behind this design choice and that your _feelings_ towards the code are changing :-)

> My God.. It's Full of Stars !

**Comment by timvisee** [comment]
Thanks a lot for your answer. This explains the situation very well. Neat! 

Personally, I always like to make the code interface as dynamic as possible for ease of use by 3rd parties, that's why I usually make objects like this mutable after construction time (in a safe manner, of course).
The change, based on your explanation is indeed obsolete now.

Also, good to see documentation being added to various parts of the project, that should help a lot while developing preventing these _issues_ in the future. Brilliant! 

Tim Visée
